### ISSUE TITLE
Support iojs v3

### ISSUE THREAD

**Issue by xzyfer** [sass/node-sass#1053]
iojs v3 was released today, this means we need to update to nan v2 which has brought some massive breaking changes.

**Comment by saper** [comment]
Working on it.

**Comment by andreimc** [comment]
:+1:

**Comment by xzyfer** [comment]
I made a start on it today but there were no end to the errors. I'm speaking at some events the next couple of days but I'll come back to this over the weekend if you don't get anywhere @saper.

**Comment by saper** [comment]
> I made a start on it today but there were no end to the errors. I'm speaking at some events the next couple of days but I'll come back to this over the weekend if you don't get anywhere @saper.

binding.cpp, create_string.cpp, custom_function_bridge.cpp are compiling fine for me now

one problem left in custom_importer_bridge.cpp

Looks not bad :)

**Comment by xzyfer** [comment]
@saper I look very forward to seeing your PR. I originally made the mistake of running the nan conversion tool. It made a real mess of things.

**Comment by kenips** [comment]
:+1: thanks for the great news @saper

**Comment by saper** [comment]
> @saper I look very forward to seeing your PR. I originally made the
> mistake of running the nan conversion tool. It made a real mess of
> things.

I did that, too, but somehow managed to fixed the damage :)

**Comment by saper** [comment]
It now compiles, just crashes :)

**Comment by saper** [comment]
I am pushing commits to #1054 as I progress (https://github.com/saper/node-sass/tree/iojs45)

Currently I am stuck at the problem with creating instances of `Color` object: 

https://travis-ci.org/sass/node-sass/jobs/74343260

````
#8  0x0000000802d492de in Nan::SetInternalFieldPointer (object=..., index=0, value=0x803419740) at ../node_modules/nan/nan.h:575
#9  0x0000000802d55a8b in SassTypes::SassValueWrapper<SassTypes::Color>::New (info=...) at ../src/sass_types/sass_value_wrapper.h:124
#10 0x0000000802d417cd in Nan::imp::FunctionCallbackWrapper (info=...) at ../node_modules/nan/nan_callbacks_12_inl.h:167
#11 0x00000000007f45a5 in v8::internal::FunctionCallbackArguments::Call (this=0x7fffffff5840, 
````
`info.This()->InternalFieldCount()` at this time is 0, which indicates an improperly constructed or deleted object.

Any help in troubleshooting this welcome :)

**Comment by saper** [comment]
node 0.10 seems to go further and crashes in `cli.js` tests:

https://travis-ci.org/sass/node-sass/jobs/74343257

What is interesting here is that we should probably have an `api.js` as well to reproduce the problem.

**Comment by saper** [comment]
Above issues seem to be resolved, feel free to test #1054

**Comment by sc0ttwad3** [comment]
If this helps.. once upgrading to iojs 3.0.0...node-sass installation is failing as others have described...but of note is installation failing to download https://github.com/sass/node-sass/releases/download/v3.2.0/win32-x64-45_binding.node ...  The releases only currently go up to **win32-x64-44_binding.node**

**Comment by saper** [comment]
For testing, one needs to check out the [modified code of node-sass in the source form](https://github.com/saper/node-sass/tree/iojs45) (for example by using `git clone --branch iojs45 https://github.com/saper/node-sass`).

Then it is necessary to compile the code from source, `npm install` will not fetch the binary as we haven't tested those changes yet.

**Comment by devinus** [comment]
What's the status of this?

**Comment by saper** [comment]
Check out and test #1054

**Comment by devinus** [comment]
Thanks @saper

**Comment by mull** [comment]
@saper @devinus what is the status of this? Unfortunately I can't participate in actually testing it but I'm looking forward to using it.

**Comment by saper** [comment]
How large is your budget?

**Comment by mull** [comment]
@saper sorry I suppose I came off as pretty greedy or so... Not a budget issue, just felt very stressed out yesterday thinking I couldn't help (but still waiting on it.) That has now changed & I'm looking into installing it and testing. Let me know if I should do anything in particular!

**Comment by mull** [comment]
Well I've tried and failed miserably at building the project. :grin:

Is there a "get started" guide somehwere? Can't see one in the wiki, readme.md or contributing.md. Sorry for the ignorance. :)

**Comment by saper** [comment]
> @saper sorry I suppose I came off as pretty greedy or so... Not
> a budget issue, just felt very stressed out yesterday thinking I
> couldn't help (but still waiting on it.) That has now changed & I'm
> looking into installing it and testing. Let me know if I should do
> anything in particular!

Just use iojs 2.x if you can't wait.

**Comment by mull** [comment]
No I can wait, and I could try and help but my attempts have been feeble so far. I appreciate the work on this and sorry again if I came off as rude.

**Comment by xzyfer** [comment]
Fixed in #1054

**Comment by xzyfer** [comment]
This will be in 3.3.0

**Comment by gajus** [comment]
@xzyfer Thank you!

**Comment by mull** [comment]
Big thank you to all involved!

**Comment by Phoenixmatrix** [comment]
What is the process/timeline for the binaries to be available via npm install? (I'm unfamiliar with node-sass release process and trying to decide if we go with iojs 2.5 or just wait a little for this to go out).

Thanks for working so hard on this!

**Comment by devinus** [comment]
@Phoenixmatrix I would recommend going with 2.5 for awhile. It seems there are several packages still not quite ready for 3.0, and going with 2.5 for another month is going to be way less headache for now.

**Comment by gajus** [comment]
@devinus Is there a reason 3.3.0 release is being delated? Currently, node-sass is the only package thats keeping me from upgrading to iojs 3.0.0, and given my rather standard stack, I am sure I am not the only one.

**Comment by mzgol** [comment]
@gajus Node.js 4.0.0 should be released in a few weeks, io.js 3 is kind of an alpha of it.

Also, note that it's already been answered in https://github.com/sass/node-sass/issues/1086#issuecomment-134548948 that a release should arrive within a few hours. But even if it's delayed a little further, it shouldn't matter too much given the imminent Node 4 release; you can always skip io.js 3 & upgrade directly to Node 4.

**Comment by gajus** [comment]
Given that several packages no longer support node at all (e.g. jsdom), going to node is not an option.

**Comment by saper** [comment]
> @devinus Is there a reason 3.3.0 release is being delated? Currently,
> node-sass is the only package thats keeping me from upgrading to iojs
> 3.0.0, and given my rather standard stack, I am sure I am not the only
> one.

Please take into account that

1) node-sass is not an ordinary npm package - we have binary
   component that has to be compatible not only with JavaScript
   and node's JS APIs, but also with C++, V8 binary interfaces
   and libuv.

2) Even io.js programmers admit that changes in V8 versions 4.4
   they are using now are far reaching and every binary
   module needs to review and modify most of their code again
   to talk to io.js properly.

The work is now mostly done, you are invited to try
git master from https://github.com/sass/node-sass for testing.

**Comment by gajus** [comment]
I am not rushing anyone. I am genuinely asking whether there are technical issues holding the release of v3.3.0. I am very grateful to everyone who contribute to this package. Thank you for the answer.

**Comment by saper** [comment]
> I am not rushing anyone. I am genuinely asking whether there are
> technical issues holding the release of v3.3.0. I am very grateful to
> everyone who contribute to this package. Thank you for the answer.

No, not at the moment, we just need testing and packaging.

**Comment by mzgol** [comment]
@gajus 
> Given that several packages no longer support node at all (e.g. jsdom), going to node is not an option.

I think you misunderstand the situation. jsdom doesn't support Node 0.10 & 0.12, it will work fine with Node 4. io.js will be dead with the release of Node 4, there will be no more io.js releases, the projects have merged.

Good news (compatibility-wise) is that Node 4 will use the same V8 version as io.js 3 so node-sass should automatically work fine with it when it starts working with io.js 3.

**Comment by gajus** [comment]
@mzgol I was not aware of that at all. Thank you for clarifying.

**Comment by thealjey** [comment]
@mzgol WOW, that does sound exciting!
However, I could not find anything about the upcoming Node.js 4.0.0 release on the internet. Could please you point me to wherever you've got this information from?
Thanks!

**Comment by saper** [comment]
> However, I could not find anything about the upcoming Node.js 4.0.0 release on the internet. Could please you point me to wherever you've got this information from?

https://medium.com/node-js-javascript/4-0-is-the-new-1-0-386597a3436d

**Comment by xzyfer** [comment]
@devinus 
>Is there a reason 3.3.0 release is being delated? 

Yes. The changes iojs 3 at the v8 were significant. This meant that we had to wait for packages we depend on to update, as well as then incorporating those changes into node-sass.

Secondly, although the new release is technically ready to go, I needed to sleep. Dropping a new release, going to bed and leaving y'all do deal with it l wasn't something I was prepared to do because I <3 you all.

>Currently, node-sass is the only package thats keeping me from upgrading to iojs 3.0.0, and given my rather standard stack, I am sure I am not the only one.

Although this may be true for you, this is not an accurate reflection of the current iojs 3 landscape.

To support iojs 3 `nan` released a new major version. `nan` is the library used by most (all?) native node extensions like filesystem watchers and websocker servers. For us this meant we had to almost completely rewrite the c++ code that binds us to LibSass.
